src/common/sys_lmdb.c

```diff
@@ -0,0 +1,511 @@
+			ldb->db_file);
+		rc = -DER_NONEXIST;
+		return rc;
+	} else if (access(ldb->db_file, R_OK | W_OK) != 0) {
+		rc = -DER_NO_PERM;
+		D_CRIT("No access to existing db file %s\n", ldb->db_file);
+		return rc;
+	}
+
+	D_DEBUG(DB_IO, "Opening %s, try_create=%d\n", ldb->db_file, try_create);
+	rc = mdb_env_create(&ldb->db_env);
+	if (rc) {
+		rc = mdb_error2daos_error(rc);
+		D_CRIT("Failed to create env handle for sysdb: "DF_RC"\n", DP_RC(rc));
+		goto out;
+	}
+
+	rc = mdb_env_open(ldb->db_env, ldb->db_file, MDB_NOSUBDIR, 0664);
+	if (rc) {
+		rc = mdb_error2daos_error(rc);
+		D_CRIT("Failed to open env handle for sysdb: "DF_RC"\n", DP_RC(rc));
+		goto out;
+	}
+
+	rc = mdb_txn_begin(ldb->db_env, NULL, 0, &ldb->db_txn);
+	if (rc) {
+		rc = mdb_error2daos_error(rc);
+		D_CRIT("Failed to begin tx for sysdb: "DF_RC"\n", DP_RC(rc));
+		goto out;
+	}
+
+	rc = mdb_dbi_open(ldb->db_txn, NULL, 0, &ldb->db_dbi);
+	if (rc) {
+		rc = mdb_error2daos_error(rc);
+		D_CRIT("Failed to open sysdb: "DF_RC"\n", DP_RC(rc));
+		goto txn_abort;
+	}
+
+
+	d_iov_set(&key, SYS_DB_MD_VER, strlen(SYS_DB_MD_VER));
+	d_iov_set(&val, &ver, sizeof(ver));
+	if (try_create) {
+		ver = SYS_DB_VERSION;
+		rc = lmm_db_upsert(db, SYS_DB_MD, &key, &val);
+		if (rc) {
+			D_CRIT("Failed to set version for sysdb: "DF_RC"\n",
+			       DP_RC(rc));
+			goto txn_abort;
+		}
+		rc = mdb_txn_commit(ldb->db_txn);
+		if (rc)
+			D_CRIT("Failed to commit version for sysdb: "DF_RC"\n",
+			       DP_RC(rc));
+		goto out;
+	} else {
+		rc = lmm_db_fetch(db, SYS_DB_MD, &key, &val);
+		if (rc) {
+			D_CRIT("Failed to read sysdb version: "DF_RC"\n",
+			       DP_RC(rc));
+			rc = -DER_INVAL;
+		}
+
+		if (ver < SYS_DB_VERSION_1 || ver > SYS_DB_VERSION)
+			rc = -DER_DF_INCOMPT;
+	}
+txn_abort:
+	mdb_txn_abort(ldb->db_txn);
+out:
+	ldb->db_txn = NULL;
+	return rc;
+}
+
+#define MAX_SMD_TABLE_LEN	32
+static int
+lmm_db_generate_key(char *table, d_iov_t *key, MDB_val *db_key)
+{
+	char	*new_key;
+	int	 table_len;
+
+	table_len = strnlen(table, MAX_SMD_TABLE_LEN);
+	db_key->mv_size = key->iov_len + table_len;
+	D_ALLOC(new_key, db_key->mv_size);
+	if (new_key == NULL)
+		return -DER_NOMEM;
+
+	memcpy(new_key, table, table_len);
+	memcpy(new_key + table_len, (char *)key->iov_buf, key->iov_len);
+	db_key->mv_data = new_key;
+
+	return 0;
+}
+
+static int
+lmm_db_unpack_key(char *table, d_iov_t *key, MDB_val *db_key)
+{
+	int 	table_len = strnlen(table, MAX_SMD_TABLE_LEN);
+	char   *buf;
+	int	len;
+
+	if (db_key->mv_size < table_len)
+		return -DER_INVAL;
+
+	len = db_key->mv_size - table_len;
+	D_ALLOC(buf, len);
+	if (buf == NULL)
+		return -DER_NOMEM;
+
+	memcpy(buf, db_key->mv_data + table_len, len);
+	d_iov_set(key, buf, len);
+
+	return 0;
+}
+
+static int
+lmm_db_fetch(struct sys_db *db, char *table, d_iov_t *key, d_iov_t *val)
+{
+	MDB_val			 db_key, db_data;
+	struct lmm_sys_db	*ldb = db2lmm(db);
+	int			 rc;
+	bool			 end_tx = false;
+
+	if (ldb->db_txn == NULL) {
```

Comment:
This could be racy if multiple threads calls lmm_db_fetch() concurrently, if you assume 'db_txn' accessing is always protected by 'lmm_db_lock()', you'd add an assert here.

BTW, does lmdb support nested transaction? If it does, we could simplify it a bit.

Reply:
It supported nested transaction, but when starting nested transaction, mdb_txn_begin must be passed parent handle, so it did not actually simplify anything even it support..